File: leetcode/186-ReverseWordsInAStringII/reverseWordsInAStringII.py
# Given an input string, reverse the string word by word. A word is defined as a sequence of non-space characters.
#
# The input string does not contain leading or trailing spaces and the words are always separated by a single space.
#
# For example,
# Given s = "the sky is blue",
# return "blue is sky the".
#
# Could you do it in-place without allocating extra space?
#
# Related problem: Rotate Array

class Solution:
    def reverseWords(self, str):
        """
        :type str: List[str]
        :rtype: void Do not return anything, modify str in-place instead.
        """
        def reverse(start, end):
            while start < end:
                str[start], str[end] = str[end], str[start]
                start += 1
                end -= 1
        # Three steps to reverse
        # 1. reverse the whole sequence
        reverse(0, len(str)-1)
        # 2. reverse each word
        start = i = 0
        while i < len(str):
            if str[i] == ' ':
                reverse(start, i-1)
                start = i+1
            i += 1
        # 3. reverse the last word
        # (b/c we check for ' ' in the loop and last word doesn't have space at the end)
        reverse(start, len(str)-1)

if __name__ == "__main__":
    sol = Solution()
    s = ["t","h","e"," ","s","k","y"," ","i","s"," ","b","l","u","e"]
    sol.reverseWords(s)
    #assert s == "blue is sky the"
    print(s)

File: leetcode/186-ReverseWordsInAStringII/reverseWordsInAStringII.cc
//    Given an input string , reverse the string word by word.
//
//    Example:
//
//    Input:  ["t","h","e"," ","s","k","y"," ","i","s"," ","b","l","u","e"]
//    Output: ["b","l","u","e"," ","i","s"," ","s","k","y"," ","t","h","e"]
//    Note:
//
//    A word is defined as a sequence of non-space characters.
//    The input string does not contain leading or trailing spaces.
//    The words are always separated by a single space.
//    Follow up: Could you do it in-place without allocating extra space?

#include <vector>
#include <cassert>

using namespace std;

class Solution {
public:
    void reverseWords(vector<char>& str) {
        // first reverse whole vector of char
        reverseHelper(str, 0, str.size()-1);
        int start = 0;
        // reverse each word
        for(int end = 0; end < str.size(); ++end)
        {
            if (str[end] == ' ' || end == str.size()-1)
            {
                if (end != str.size()-1)
                {
                    reverseHelper(str, start, end-1);
                }
                else
                {
                    reverseHelper(str, start, end);
                }
                start = end + 1;
            }
        }
    }
private:
    void reverseHelper(vector<char>& str, int start, int end)
    {
        int i = start;
        int j = end;
        while (i < j)
        {
            swap(str[i], str[j]);
            i++;
            j--;
        }
    }
};

using ptr2reverseWords = void (Solution::*)(vector<char>&);

void test(ptr2reverseWords pfcn)
{
    Solution sol;
    vector<char> input = {'t','h','e',' ','s','k','y',' ','i','s',' ','b','l','u','e'};
    vector<char> ans = {'b','l','u','e',' ','i','s',' ','s','k','y',' ','t','h','e'};
    (sol.*pfcn)(input);
    assert( input == ans);
}

int main()
{
    ptr2reverseWords pfcn = &Solution::reverseWords;
    test(pfcn);
}